imapclient/notify.go

```diff
@@ -0,0 +1,154 @@
+package imapclient
+
+import (
+	"sync/atomic"
+
+	"github.com/emersion/go-imap/v2"
+	"github.com/emersion/go-imap/v2/internal/imapwire"
+)
+
+// Notify sends a NOTIFY command (RFC 5465).
+//
+// The NOTIFY command allows clients to request server-push notifications
+// for mailbox events like new messages, expunges, flag changes, etc.
+//
+// When NOTIFY SET is active, the server may send unsolicited responses at any
+// time (STATUS, FETCH, EXPUNGE, LIST responses). These unsolicited responses
+// are delivered via the UnilateralDataHandler callbacks set in
+// imapclient.Options.
+//
+// When the server sends an untagged OK [NOTIFICATIONOVERFLOW] response, the
+// Overflow() channel on the returned NotifyCommand will be closed. This
+// indicates the server has disabled all notifications and the client should
+// re-issue the NOTIFY command if needed.
+//
+// This requires support for the NOTIFY extension.
+func (c *Client) Notify(options *imap.NotifyOptions) (*NotifyCommand, error) {
+	cmd := &NotifyCommand{
+		options:  options,
+		overflow: make(chan struct{}),
+	}
+	enc := c.beginCommand("NOTIFY", cmd)
+	encodeNotifyOptions(enc.Encoder, options)
+	enc.end()
+
+	if err := cmd.Wait(); err != nil {
+		return nil, err
+	}
+
+	// Start monitoring for connection failures
+	// Only needed if NOTIFY SET was sent (not NOTIFY NONE)
+	if options != nil && len(options.Items) > 0 {
+		go cmd.monitorConnection(c)
+	}
+
+	return cmd, nil
+}
+
+// encodeNotifyOptions encodes NOTIFY command options to the encoder.
+func encodeNotifyOptions(enc *imapwire.Encoder, options *imap.NotifyOptions) {
+	if options == nil || len(options.Items) == 0 {
+		// NOTIFY NONE - disable all notifications
+		enc.SP().Atom("NONE")
+	} else {
+		// NOTIFY SET
+		enc.SP().Atom("SET")
+
+		if options.STATUS {
+			enc.SP().List(1, func(i int) {
+				enc.Atom("STATUS")
+			})
+		}
+
+		// Encode each notify item
+		for _, item := range options.Items {
+			// Validate the item before encoding
+			if item.MailboxSpec == "" && len(item.Mailboxes) == 0 {
+				// Skip invalid items - this shouldn't happen with properly constructed NotifyOptions
+				continue
```

Comment:
We should probably return an error here?

Reply:
I can't find any other method which uses an encoder and returns an error.

Is it fine to leave the command half-written in this case? (this indicates a programmer error anyway).

Reply:
For some commands we `panic` when fed an invalid state (e.g. see `GetMetadata`). For some commands we validate upfront and we return a command struct already completed with an error (e.g. see `Enable`). Latter is more complicated and better suited for cases where the failure doesn't indicate a programmer error (`Enable` is probably a bad usage for this).

Reply:
Do you prefer for this to panic or return an error?

Reply:
It seems like this comment has been addressed?